packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -143,11 +154,30 @@ export class BlobManager {
             values.map((entry) => detached ? this.detachedBlobIds.add(entry) : this.blobIds.add(entry));
         }
         this.logger.sendTelemetryEvent({ eventName: "AttachmentBlobsLoaded", count });
+        if (redirectTable) {
+            this.redirectTable = new Map(redirectTable);
+            this.logger.sendTelemetryEvent({ eventName: "BlobRedirectTableLoaded", count: redirectTable.length });
+        }
     }
 
-    public snapshot(): ITree {
-        const blobIds = this.runtime.attachState === AttachState.Detached ? this.detachedBlobIds : this.blobIds;
-        const entries = [...blobIds].map((id) => new AttachmentTreeEntry(id, id));
-        return { entries };
+    public snapshot(): { ids: ITree, table?: string } {
+        // If we have a redirect table it means the container is about to transition to "Attaching" state, so we need
+        // to return an actual snapshot containing all the real storage IDs we know about.
+        const attachingOrAttached = !!this.redirectTable || this.runtime.attachState !== AttachState.Detached;
+        const blobIds = attachingOrAttached ? this.blobIds : this.detachedBlobIds;
+        const entries: ITreeEntry[] = [...blobIds].map((id) => new AttachmentTreeEntry(id, id));
+        return {
+            ids: { entries },
+            table: this.redirectTable ? JSON.stringify(Array.from(this.redirectTable.entries())) : undefined,
+        };
+    }
+
+    public setRedirectTable(table: Map<string, string>) {
+        for (const [localId, storageId] of table) {
```

Comment:
I believe this API should be allowed to be called only once and only in specific state of container. Can we assert all these pre-conditions?